pkg/kv/kvserver/loqrecovery/collect.go

```diff
@@ -62,3 +76,68 @@ func CollectReplicaInfo(
 	}
 	return loqrecoverypb.NodeReplicaInfo{Replicas: replicas}, nil
 }
+
+// checkUncommittedDescriptorsInRaftLog iterates over committed but not applied
+// part of raft log and searches for writes to range descriptor. If any value
+// referencing local range descriptor key is found then we have a potential
+// pending update.
+func checkUncommittedDescriptorsInRaftLog(
+	rangeID roachpb.RangeID,
+	rangeKey roachpb.RKey,
+	startIndex uint64,
+	endIndex uint64,
+	reader storage.Reader,
+) (bool, error) {
+	rangeDescKey := keys.RangeDescriptorKey(rangeKey)
+
+	key := keys.RaftLogKey(rangeID, startIndex)
+	endKey := keys.RaftLogKey(rangeID, endIndex)
+	iter := reader.NewMVCCIterator(storage.MVCCKeyIterKind, storage.IterOptions{
+		UpperBound: endKey,
+	})
+	defer iter.Close()
+
+	var meta enginepb.MVCCMetadata
+	var ent raftpb.Entry
+
+	iter.SeekGE(storage.MakeMVCCMetadataKey(key))
+	for ; ; iter.Next() {
+		if ok, err := iter.Valid(); err != nil || !ok {
+			return false, err
+		}
+		if err := protoutil.Unmarshal(iter.UnsafeValue(), &meta); err != nil {
+			return false, errors.Wrap(err, "unable to decode raft log MVCCMetadata")
+		}
+		if err := storage.MakeValue(meta).GetProto(&ent); err != nil {
+			return false, errors.Wrap(err, "unable to unmarshal raft Entry")
+		}
+		if ent.Type != raftpb.EntryNormal || len(ent.Data) == 0 {
+			continue
+		}
+		_, cmdData := kvserver.DecodeRaftCommand(ent.Data)
+		if len(cmdData) == 0 {
+			continue
+		}
+		var raftCmd kvserverpb.RaftCommand
+		if err := protoutil.Unmarshal(cmdData, &raftCmd); err != nil {
+			return false, errors.Wrap(err, "unable to unmarshal raft command")
+		}
+		batchReader, err := storage.NewRocksDBBatchReader(raftCmd.WriteBatch.Data)
+		if err != nil {
+			return false, errors.Wrap(err, "unable to read entries from raft command")
+		}
+		for batchReader.Next() {
+			switch batchReader.BatchType() {
+			case storage.BatchTypeValue:
+				mvccKey, err := batchReader.MVCCKey()
+				if err == nil {
+					if rangeDescKey.Equal(mvccKey.Key) {
```

Comment:
The operation that performs the change is really the commit of such an operation, so I think you'd be looking for operations that perform an engine clear on the range descriptor key meta record, where the clear originates from this line during evaluation:

https://github.com/cockroachdb/cockroach/blob/c6dc23e9b1978e7ce0f965dd6e033c02e306a3f3/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L483-L482

which calls down to the actual clear here:

https://github.com/cockroachdb/cockroach/blob/52c2df4fb00d943328dc59df9dc210a40a556ab2/pkg/storage/mvcc.go#L3147

The distinction is important because the code here doesn't handle the case in which the intent writes are committed but the commit is not:

```
Put(txn, descKey) <-- applied
Something else    <-- committed
Commit(txn)
```

In this case the above code doesn't find anything or maybe it does, but if so I'm not sure if intentionally so.

There's also existing code (I remember that from unsafe-remove-dead-replicas) that aborts intents on the range descriptor; this should somehow play into it here since it's trying to handle similar issues (though more shallowly). I think you want to check

- is there an intent on the range desc (as of applied index)
- if not, is there a write in the raft log that will write such an intent? (I think that's your code!)
- if either of the above, check if the intent could possibly have committed (that's my suggestion above: is there a corresponding EndTxn in the tail of the log)
- if so, bail
- if not, we're fine: the txn will ultimately be aborted.

As always there is the residual risk that the txn actually did commit but we never learned about the corresponding log entry. This information gap can't be bridged unless we also add offline analysis of meta2 into the equation, which we can't right now.

Reply:
I changed it to bail on any sequence other than a single clear lock for descriptor. We currently drop the intent if it is found. If we want to keep transaction then we might need to apply it as well as that implies other ranges has changed as well. So in those cases we'll end up with incomplete coverage.